src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs

```diff
@@ -13,45 +13,58 @@ internal class PhysicalProjectTreeStorage : IPhysicalProjectTreeStorage
     {
         private readonly Lazy<IFileSystem> _fileSystem;
         private readonly ActiveConfiguredProject<IFolderManager> _folderManager;
-        private readonly Lazy<IProjectTreeService> _treeService;
-        private readonly Lazy<IProjectTreeProvider> _treeProvider;
+        private readonly ActiveConfiguredProject<IProjectItemProvider> _sourceItemProvider;
+        private readonly IProjectTreeService _treeService;
         private readonly UnconfiguredProject _unconfiguredProject;
 
         [ImportingConstructor]
-        public PhysicalProjectTreeStorage([Import(ExportContractNames.ProjectTreeProviders.PhysicalProjectTreeService)]Lazy<IProjectTreeService> treeService,
-                                          [Import(ExportContractNames.ProjectTreeProviders.PhysicalViewTree)]Lazy<IProjectTreeProvider> treeProvider,
+        public PhysicalProjectTreeStorage([Import(ExportContractNames.ProjectTreeProviders.PhysicalProjectTreeService)]IProjectTreeService treeService,
                                           Lazy<IFileSystem> fileSystem,
                                           ActiveConfiguredProject<IFolderManager> folderManager,
+                                          [Import(ExportContractNames.ProjectItemProviders.SourceFiles)]ActiveConfiguredProject<IProjectItemProvider> sourceItemProvider,
                                           UnconfiguredProject project)
         {
             _treeService = treeService;
-            _treeProvider = treeProvider;
+            _sourceItemProvider = sourceItemProvider;
             _fileSystem = fileSystem;
             _folderManager = folderManager;
             _unconfiguredProject = project;
         }
 
-        public Task<IProjectTree> CreateFolderAsync(string path)
+        public async Task<IProjectItem?> CreateEmpyFileAsync(string path)
         {
             Requires.NotNullOrEmpty(path, nameof(path));
 
-            if (_treeService.Value.CurrentTree == null)
-                throw new InvalidOperationException("Physical project tree has not yet been published.");
+            return await AddToProjectAsync(path, waitForFileSystemUpdates: true, async fullPath =>
+            {
+                using (_fileSystem.Value.Create(fullPath)) { }
 
-            string fullPath = _unconfiguredProject.MakeRooted(path);
+                await _sourceItemProvider.Value.AddAsync(fullPath);
+
+            }) as IProjectItem;
+        }
+
+        public Task<IProjectTree?> CreateFolderAsync(string path)
+        {
+            Requires.NotNullOrEmpty(path, nameof(path));
 
-            return AddToProjectAsync(fullPath);
+            return AddToProjectAsync(path, waitForFileSystemUpdates: false, async fullPath =>
```

Comment:
`CreateEmpyFileAsync` awaits the `AddToProjectAsync` task, whereas this returns it directly. Perhaps they should be consistent. Awaiting in both will create clearer stack traces.

Reply:
Not convinced by this - one needs the result, one doesn't, and would just add unneeded state machinary.

Reply:
Ah, I missed the `as` cast at the end.